db/db_impl/db_impl_write.cc

```diff
@@ -1869,27 +1868,26 @@ Status DBImpl::DelayWrite(uint64_t num_bytes, WriteThread& write_thread,
   return s;
 }
 
-// REQUIRES: mutex_ is held
 // REQUIRES: this thread is currently at the front of the writer queue
-void DBImpl::WriteBufferManagerStallWrites() {
-  mutex_.AssertHeld();
-  // First block future writer threads who want to add themselves to the queue
-  // of WriteThread.
-  write_thread_.BeginWriteStall();
-  mutex_.Unlock();
-
-  // Change the state to State::Blocked.
-  static_cast<WBMStallInterface*>(wbm_stall_.get())
-      ->SetState(WBMStallInterface::State::BLOCKED);
-  // Then WriteBufferManager will add DB instance to its queue
-  // and block this thread by calling WBMStallInterface::Block().
-  write_buffer_manager_->BeginWriteStall(wbm_stall_.get());
-  wbm_stall_->Block();
-
-  mutex_.Lock();
-  // Stall has ended. Signal writer threads so that they can add
-  // themselves to the WriteThread queue for writes.
-  write_thread_.EndWriteStall();
+void DBImpl::MaybeWriteBufferManagerStallWrites() {
+  bool begin_stall =
+      write_buffer_manager_->MaybeBeginWriteStall(wbm_stall_.get());
+  if (begin_stall) {
```

Comment:
Optimization 1: 

Block future writers only if `begin_stall` is true here. That is because we could have runtime-changed the parameters (e.g, allow_stall_, buffer_size_) that affects stall conditions between `if (UNLIKELY(status.ok() && write_buffer_manager_->ShouldStall())) {` and  `bool begin_stall = write_buffer_manager_->MaybeBeginWriteStall(wbm_stall_.get());`

With this, I'm also moving the responsibility of setting StallInterface's state `static_cast<WBMStallInterface*>(wbm_stall_.get())
->SetState(WBMStallInterface::State::BLOCKED);` into `WBM::MaybeBeginWriteStall()` if we actually need to stall (that is `begin_stall` is returned as true).

Concern 1: Is it concerning that now call `write_buffer_manager_->MaybeBeginWriteStall` without `write_thread_.BeginWriteStall();` first?

Reply:
> Concern 1: Is it concerning that we now call write_buffer_manager_->MaybeBeginWriteStall without write_thread_.BeginWriteStall(); first?

Maybe I've found myself an answer to this question - it doesn't make sense to me to share unique ptr `wbm_stall_` across multiple threads which can happen now without holding lock nor stalling incoming write threads. Maybe a better impl should be the following:

```
  if (UNLIKELY(status.ok() && write_buffer_manager_->ShouldStall())) {
    if (write_options.no_slowdown) {
      status = Status::Incomplete("Write stall");
    } else {
      InstrumentedMutexLock l(db_mutex_);
      MaybeWriteBufferManagerStallWrites();
    }
  }
 ```
 
 ```
void DBImpl::MaybeWriteBufferManagerStallWrites() {
  bool begin_stall =
      write_buffer_manager_->MaybeBeginWriteStall(wbm_stall_.get());
  if (begin_stall) {
    // mutex_.Lock();
    // Block future writer threads who want to add themselves to the queue
    // of WriteThread.
    write_thread_.BeginWriteStall();
    mutex_.Unlock();
    // Block this thread by calling WBMStallInterface::Block().
    wbm_stall_->Block();
    mutex_.Lock();
    // Stall has ended. Signal writer threads so that they can add
    // themselves to the WriteThread queue for writes.
    write_thread_.EndWriteStall();
  } else {
    return;
  }
}
 ```
 
But then now we grab the lock instead of stop incoming writer threads when we are `WBM::MaybeBeginWriteStall()` - is this really better than the original one where we stop incoming writer threads when we are `WBM::BeginWriteStall()`??

Reply:
As discussed with @akankshamahajan15, we will revert this optimization as it's difficult to verify it's correct and not needed at this moment.